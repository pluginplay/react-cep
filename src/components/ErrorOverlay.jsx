import React from 'react'
import { observer } from 'mobx-react'
import PropTypes from 'prop-types'
import Overlay from './Overlay'
import OverlayContent from './OverlayContent'
import ColorHeader from './ColorHeader'
import MonospaceContent from './MonospaceContent'
import styled from 'styled-components'

@observer
class ErrorOverlay extends React.Component {
  constructor (props) {
    super(props)
    this._onCloseClickedBound = this.onCloseClicked.bind(this)
    this._onWhatMeanClickedBound = this.onWhatMeanClicked.bind(this)
    this._onReportClickedBound = this.onReportClicked.bind(this)
  }

  onWhatMeanClicked (e) {
    if (e) {
      e.preventDefault()
    }
    this.props.systemState.showConfirmation('When you send the developers an error report, you send very useful ' +
      'information about what went wrong that allows us to improve the plugin. The error report will contain ' +
      'your AE version, installed extensions, and what operating system you\'re running.')
  }

  onReportClicked (e) {
    if (e) {
      e.preventDefault()
    }
    if (this.props.onReportClicked) {
      this.props.onReportClicked()
    }
    this.props.systemState.hideErrorMessage()
  }

  onCloseClicked () {
    this.props.systemState.hideErrorMessage()
  }

  render () {
    return (
      <Overlay visible={this.props.systemState.showErrorOverlay} column full className={this.props.className}>
        <ColorHeader color={'red'}>
          <span>Oh dear, an error!</span>
          <a href={'#'} onClick={this._onCloseClickedBound}>Close</a>
        </ColorHeader>
        <OverlayContent full column>
          <p>
            Looks like we've run into a problem. You could help the developers out and <a
              href={'#'} className={'send-button'} onClick={this._onReportClickedBound}>send the error report.</a>
            <br /><a href={'#'} onClick={this._onWhatMeanClickedBound}>What does this mean?</a>
          </p>
          <MonospaceContent grow={1}>{this.props.systemState.errorMessage}</MonospaceContent>
        </OverlayContent>
      </Overlay>
    )
  }
}

ErrorOverlay.propTypes = {
  systemState: PropTypes.object.isRequired,
  onReportClicked: PropTypes.func
}

export default styled(ErrorOverlay)`
  .send-button {
    background: ${props => props.theme.ae.systemHighlightColor};
    padding: 1px 2px;
    color: white;
    font-weight: bold;
    white-space: nowrap;
  }
`
