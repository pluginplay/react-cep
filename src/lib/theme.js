import * as csInterface from 'cep-lib/csinterface'
import { lighten, darken } from 'polished'

export const getDefaultTheme = () => {
  const ae = csInterface.getThemeInformation()
  ae.systemHighlightColor = '#0078d7' // We have to hard-code this because Windows gets the wrong color.
  return {
    ae,
    background: {
      lighter: lighten(0.2, ae.panelBackgroundColor),
      light: lighten(0.1, ae.panelBackgroundColor),
      dark: darken(0.05, ae.panelBackgroundColor),
      darker: darken(0.1, ae.panelBackgroundColor)
    },
    colors: {
      green: '#23EB87',
      red: '#eb093f',
      blue: ae.systemHighlightColor
    }
  }
}
