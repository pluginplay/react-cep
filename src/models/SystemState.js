import { observable, action } from 'mobx'

export default class SystemState {
  @observable progress = 0
  @observable progressLabel = 'Loading'
  @observable showProgress = false
  @observable showProgressComplete = false
  @observable errorMessage = null
  @observable errorObject = null
  @observable showErrorOverlay = false
  @observable debugMessage = null
  @observable showDebugOverlay = null
  @observable confirmationMessage = null
  @observable confirmationCallback = null
  @observable showConfirmationOverlay = false

  @action setProgress (progress, label, showProgress = true) {
    this.showProgressComplete = false
    this.showProgress = showProgress
    this.progressLabel = label
    this.progress = progress
  }

  @action progressComplete () {
    this.progress = 100
    this.showProgressComplete = true
    setTimeout(() => {
      this.hideProgress()
    }, 3000)
  }

  @action hideProgress () {
    this.showProgress = false
  }

  @action setErrorMessage (errorText, error) {
    this.errorMessage = errorText
    this.errorObject = error
    this.showErrorOverlay = true
  }

  @action hideErrorMessage () {
    this.errorMessage = ''
    this.errorObject = null
    this.showErrorOverlay = false
  }

  @action addDebugMessage (message) {
    console.log(`[DEBUG] ${message}`)
    if (this.debugMessage === null) {
      this.debugMessage = message
    } else {
      this.debugMessage += `\n${message}`
    }
  }

  @action showDebug () {
    this.showDebugOverlay = true
  }

  @action clearDebug () {
    this.showDebugOverlay = false
    this.debugMessage = null
  }

  @action showConfirmation (message, callback) {
    this.showConfirmationOverlay = true
    this.confirmationMessage = message
    this.confirmationCallback = callback
  }

  @action hideConfirmation () {
    this.showConfirmationOverlay = false
    this.confirmationMessage = null
    this.confirmationCallback = null
  }
}
