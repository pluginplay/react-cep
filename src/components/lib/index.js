import SystemState from '../../models/SystemState'
import PersistedState from '../../models/PersistedState'

export { SystemState, PersistedState }
