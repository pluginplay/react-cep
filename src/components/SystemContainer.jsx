import React from 'react'
import * as csInterface from 'cep-lib/csinterface'
import { ThemeProvider } from 'styled-components'
import { observer } from 'mobx-react'
import Flex from 'styled-flex-component'
import PropTypes from 'prop-types'
import Panel from './Panel'
import ProgressOverlay from './ProgressOverlay'
import ErrorOverlay from './ErrorOverlay'
import DebugOverlay from './DebugOverlay'
import ConfirmationOverlay from './ConfirmationOverlay'
import SimpleMessageOverlay from './SimpleMessageOverlay'
import { getDefaultTheme } from '../lib/theme'

@observer
class SystemContainer extends React.Component {
  constructor (props) {
    super(props)
    const _onErrorBound = this.onError.bind(this)
    csInterface.addEventListener(props.errorEvent, _onErrorBound)
    csInterface.addEventListener(props.debugEvent, this.onDebug.bind(this))
    window.addEventListener('unhandledrejection', _onErrorBound)
    window.addEventListener('error', _onErrorBound)
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.errorEvent !== this.props.errorEvent) {
      csInterface.addEventListener(nextProps.errorEvent, this.onError.bind(this))
    }
    if (nextProps.debugEvent !== this.props.debugEvent) {
      csInterface.addEventListener(nextProps.debugEvent, this.onDebug.bind(this))
    }
  }

  onDebug (event) {
    this.props.systemState.addDebugMessage(event.data)
  }

  onError (event) {
    this.props.systemState.hideProgress()
    let error = null
    if (event.error && event.error.stack) {
      event = event.error
    }
    if (event.data) {
      error = JSON.stringify(event.data)
    } else if (event.reason && event.reason.stack) {
      error = event.reason.stack
    } else if (event.stack) {
      error = event.stack
    } else if (event.message) {
      error = event.message
    } else {
      error = JSON.stringify(event)
    }
    if (error.indexOf('Allow Scripts to Write Files and Access Network') !== -1) {
      this.props.systemState.showConfirmation(`
        Uh oh! Since this plugin writes files to the disk, you'll need to make sure to
        enable the "Allow Scripts to Write Files and Access Network" setting inside
        General preferences (Preferences > General). Once you do that, try the action
        again.
      `.trim())
      return
    }
    if (this.props.filterErrorEvent && this.props.filterErrorEvent(error, event) === false) {
      return
    }
    this.props.systemState.setErrorMessage(error.replace(/\\n/g, '\n'), event)
    if (this.props.systemState.debugModeEnabled) {
      this.props.systemState.showDebug()
    }
  }

  render () {
    const theme = Object.assign(getDefaultTheme(), this.props.theme)
    return (
      <ThemeProvider theme={theme}>
        <Panel className={this.props.className}>
          <DebugOverlay systemState={this.props.systemState} />
          <ErrorOverlay systemState={this.props.systemState} onReportClicked={this.props.onReportClicked} />
          <ProgressOverlay systemState={this.props.systemState} />
          <ConfirmationOverlay systemState={this.props.systemState} />
          <SimpleMessageOverlay systemState={this.props.systemState} />
          <Flex full column>
            {this.props.children}
          </Flex>
        </Panel>
      </ThemeProvider>
    )
  }
}

SystemContainer.propTypes = {
  systemState: PropTypes.object.isRequired,
  filterErrorEvent: PropTypes.func,
  onReportClicked: PropTypes.func,
  errorEvent: PropTypes.string.isRequired,
  debugEvent: PropTypes.string.isRequired,
  theme: PropTypes.object.isRequired
}

SystemContainer.defaultProps = {
  theme: {}
}

export default SystemContainer
