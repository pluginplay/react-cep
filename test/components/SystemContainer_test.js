import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { expect } from '../setup'
import SystemState from '../../src/components/SystemState'
import SystemContainer from '../../src/components/SystemContainer'
import { ThemeProvider } from 'styled-components'
import { getDefaultTheme } from '../../src/lib/theme'
import * as csInterface from 'cep-lib/csinterface'
import ErrorOverlay from '../../src/components/ErrorOverlay'

let showConfirmationSpy = null
let setErrorMessageSpy = null
let addDebugMessageSpy = null
let showDebugSpy = null
let hideProgressSpy = null
let addEventListenerMock = null
let systemState = null

jest.mock('cep-lib/csinterface')

beforeEach(() => {
  showConfirmationSpy = jest.spyOn(SystemState.prototype, 'showConfirmation')
  setErrorMessageSpy = jest.spyOn(SystemState.prototype, 'setErrorMessage')
  showDebugSpy = jest.spyOn(SystemState.prototype, 'showDebug')
  addDebugMessageSpy = jest.spyOn(SystemState.prototype, 'addDebugMessage')
  hideProgressSpy = jest.spyOn(SystemState.prototype, 'hideProgress')
  addEventListenerMock = jest.fn()
  csInterface.getThemeInformation.mockReturnValue({
    baseFontsize: 10,
    baseFontFamily: 'Adobe Clean',
    systemHighlightColor: '#FFFFFF',
    appBarBackgroundColor: '#FFFFFF',
    panelBackgroundColor: '#FFFFFF'
  })
  window.addEventListener = addEventListenerMock
  systemState = new SystemState()
})

afterEach(() => {
  jest.resetAllMocks()
  window.addEventListener = () => {}
})

it('renders properly', () => {
  const tree = renderer
    .create(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'}>
      <p>foobar</p></SystemContainer>)
    .toJSON()
  jestExpect(tree).toMatchSnapshot()
})

it('filters errors when they appear', () => {
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  const errorCallback = addEventListenerMock.mock.calls[0][1]
  errorCallback({ data: { foo: 'bar' } })
  errorCallback({ message: 'foobar' })
  errorCallback({ stack: 'foobartwo' })
  errorCallback({ bleh: 'bab' })
  errorCallback({
    error: {
      stack: 'stackone',
      message: 'foobarthree'
    }
  })
  errorCallback({
    reason: {
      stack: 'stacktwo'
    }
  })
  expect(setErrorMessageSpy.mock.calls[0]).to.deep.equal([
    JSON.stringify({ foo: 'bar' }),
    { data: { foo: 'bar' } }
  ])
  expect(setErrorMessageSpy.mock.calls[1][0]).to.equal('foobar')
  expect(setErrorMessageSpy.mock.calls[2][0]).to.equal('foobartwo')
  expect(setErrorMessageSpy.mock.calls[3][0]).to.equal(JSON.stringify({ bleh: 'bab' }))
  expect(setErrorMessageSpy.mock.calls[4][0]).to.equal('stackone')
  expect(setErrorMessageSpy.mock.calls[5][0]).to.equal('stacktwo')
  expect(setErrorMessageSpy.mock.calls.length).to.equal(6)
})

it('forwards the onReportClicked prop to ErrorOverlay', () => {
  const callback = () => false
  const wrapper = shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'}
    onReportClicked={callback} debugEvent={'foo.debug'} />)
  expect(wrapper.find(ErrorOverlay).first().prop('onReportClicked')).to.equal(callback)
})

it('hides progress when an error occurs', () => {
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  const errorCallback = addEventListenerMock.mock.calls[0][1]
  errorCallback({ message: 'foo' })
  expect(hideProgressSpy.mock.calls).to.have.length(1)
})

it('shows a confirmation when the error is about allowing scripts to access network', () => {
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  const errorCallback = addEventListenerMock.mock.calls[0][1]
  errorCallback({ message: 'bleh Allow Scripts to Write Files and Access Network bleh' })
  expect(showConfirmationSpy.mock.calls).to.have.length(1)
  expect(showConfirmationSpy.mock.calls[0][0]).to.be.a('string')
})

it('shows debug when configured and an error occurs', () => {
  systemState.debugModeEnabled = true
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  const errorCallback = addEventListenerMock.mock.calls[0][1]
  errorCallback({ message: 'poo' })
  expect(showDebugSpy.mock.calls).to.have.length(1)
})

it('does not show debug when configured and an error occurs', () => {
  systemState.debugModeEnabled = false
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  const errorCallback = addEventListenerMock.mock.calls[0][1]
  errorCallback({ message: 'poo' })
  expect(showDebugSpy.mock.calls).to.have.length(0)
})

it('filters an error with the custom callback properly', () => {
  const filterEvent = (error, event) => {
    expect(error).to.equal('foobar')
    expect(event).to.deep.equal({ message: 'foobar' })
    return false
  }
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'}
    filterErrorEvent={filterEvent} />)
  const errorCallback = addEventListenerMock.mock.calls[0][1]
  errorCallback({ message: 'foobar' })
  expect(setErrorMessageSpy.mock.calls).to.have.length(0)
})

it('registers the error handler with CSInterface', () => {
  csInterface.addEventListener.mockReturnValueOnce(null)
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  expect(csInterface.addEventListener.mock.calls[0][0]).to.equal('foo.error')
  expect(csInterface.addEventListener.mock.calls[0][1]).to.be.a('function')
})

it('registers the debug handler with CSInterface', () => {
  csInterface.addEventListener.mockReturnValueOnce(null)
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  expect(csInterface.addEventListener.mock.calls[1][0]).to.equal('foo.debug')
  expect(csInterface.addEventListener.mock.calls[1][1]).to.be.a('function')
})

it('registers the error handler with unhandledrejection', () => {
  shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'} />)
  expect(addEventListenerMock.mock.calls[0][0]).to.equal('unhandledrejection')
  expect(addEventListenerMock.mock.calls[0][1]).to.be.a('function')
})

it('re-registers the new error handler with CSInterface upon change', () => {
  csInterface.addEventListener.mockReturnValue(null)
  const wrapper = shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'}
    debugEvent={'foo.debug'} />)
  wrapper.setProps({ errorEvent: 'foo.error.two' })
  expect(csInterface.addEventListener.mock.calls[0][0]).to.equal('foo.error')
  expect(csInterface.addEventListener.mock.calls[0][1]).to.be.a('function')
  expect(csInterface.addEventListener.mock.calls[2][0]).to.equal('foo.error.two')
  expect(csInterface.addEventListener.mock.calls[2][1]).to.be.a('function')
})

it('re-registers the new debug handler with CSInterface upon change', () => {
  csInterface.addEventListener.mockReturnValue(null)
  const wrapper = shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'}
    debugEvent={'foo.debug'} />)
  wrapper.setProps({ debugEvent: 'foo.debug.two' })
  expect(csInterface.addEventListener.mock.calls[1][0]).to.equal('foo.debug')
  expect(csInterface.addEventListener.mock.calls[1][1]).to.be.a('function')
  expect(csInterface.addEventListener.mock.calls[2][0]).to.equal('foo.debug.two')
  expect(csInterface.addEventListener.mock.calls[2][1]).to.be.a('function')
})

it('merges the theme', () => {
  const wrapper = shallow(<SystemContainer systemState={systemState} errorEvent={'foo.error'} debugEvent={'foo.debug'}
    theme={{foo: 'bar'}}/>)
  expect(wrapper.find(ThemeProvider).first()).to.have.prop('theme').deep.equal({
    ...(getDefaultTheme()),
    foo: 'bar'
  })
})
