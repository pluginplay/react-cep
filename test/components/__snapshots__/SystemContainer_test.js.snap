// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
}

.c5 {
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  display: block;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c0 {
  background: #FFFFFF;
  color: white;
  font-family: "Adobe Clean","Segoe UI","San Francisco",sans-serif;
  font-size: pt;
  position: relative;
}

.c9 {
  padding: 4px;
  border-radius: 10px;
  background: #e6e6e6;
  margin: 10px auto;
  max-width: 80%;
  width: 100%;
}

.c10 {
  background: #0078d7;
  border-radius: 5px;
  max-width: 100%;
  height: 6px;
  width: 0%;
  -webkit-transition: width 0.5s ease-in-out,background 0.5s linear;
  transition: width 0.5s ease-in-out,background 0.5s linear;
  position: relative;
  overflow: hidden;
}

.c10:before {
  display: block;
  content: ' ';
  background: linear-gradient(to right,rgba(255,255,255,0) 0%,rgba(255,255,255,0.35) 50%,rgba(255,255,255,0) 100%);
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  width: 100px;
  -webkit-animation: gdzmdV 2s ease-in-out infinite;
  animation: gdzmdV 2s ease-in-out infinite;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  -webkit-transition: opacity 0.25s linear,z-index 0s linear 0.25s;
  transition: opacity 0.25s linear,z-index 0s linear 0.25s;
  text-align: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: #f2f2f2;
  z-index: -1;
  opacity: 0;
}

.c1 a {
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  opacity: 0.75;
  -webkit-transition: opacity 0.2s linear;
  transition: opacity 0.2s linear;
}

.c1 a:hover {
  opacity: 1;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: opacity 0.25s linear,z-index 0s linear 0.25s;
  transition: opacity 0.25s linear,z-index 0s linear 0.25s;
  text-align: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: #f2f2f2;
  z-index: -1;
  opacity: 0;
}

.c8 a {
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  opacity: 0.75;
  -webkit-transition: opacity 0.2s linear;
  transition: opacity 0.2s linear;
}

.c8 a:hover {
  opacity: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  padding: 10px;
  box-sizing: border-box;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 10px;
  box-sizing: border-box;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  background: #0078d7;
  color: white;
  padding: 10px;
  font-size: 16px;
  font-weight: bold;
  line-height: 1;
}

.c2 span {
  margin: 0 auto;
}

.c2 a {
  margin: 0 3px 0 auto;
  font-size: 14px;
  font-weight: normal;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  background: #eb093f;
  color: white;
  padding: 10px;
  font-size: 16px;
  font-weight: bold;
  line-height: 1;
}

.c7 span {
  margin: 0 auto;
}

.c7 a {
  margin: 0 3px 0 auto;
  font-size: 14px;
  font-weight: normal;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 {
  background: #e6e6e6;
  border-radius: 2px;
  padding: 5px;
  margin: 0;
  width: 100%;
  box-sizing: border-box;
  font-family: monospace;
  overflow: auto;
  text-align: left;
}

.c6 .send-button {
  background: #0078d7;
  padding: 1px 2px;
  color: white;
  font-weight: bold;
  white-space: nowrap;
}

.c12 {
  padding: 4px 8px;
  margin: 10px auto;
  min-width: 100px;
  font-family: "Adobe Clean","Segoe UI","San Francisco",sans-serif;
  color: white;
  background: #f2f2f2;
  border: solid 1px #fff;
  font-size: pt;
  outline: none;
  box-sizing: border-box;
  -webkit-transition: background 0.1s ease-in-out,border-color 0.1s ease-in-out,-webkit-transform 0.1s ease-in-out;
  -webkit-transition: background 0.1s ease-in-out,border-color 0.1s ease-in-out,transform 0.1s ease-in-out;
  transition: background 0.1s ease-in-out,border-color 0.1s ease-in-out,transform 0.1s ease-in-out;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  cursor: pointer;
  display: inline-block;
}

.c12:hover,
.c12:focus,
.c12:active {
  border-color: #0078d7;
  background: #e6e6e6;
}

.c12:active {
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
  background: #FFFFFF;
}

.c13 {
  background: none;
  -webkit-transition: opacity 0.25s linear,z-index 0s linear 0.25s;
  transition: opacity 0.25s linear,z-index 0s linear 0.25s;
}

.c13 > p {
  padding: 10px 15px;
  background: #e6e6e6;
  border: solid 1px #fff;
  box-shadow: 0 0 3px #f2f2f2;
  border-radius: 2px;
}

<div
  className="c0"
  id="panel"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      color="blue"
    >
      <span>
        Debug Information
      </span>
      <a
        href="#"
        onClick={[Function]}
      >
        Close
      </a>
    </div>
    <div
      className="c3"
    >
      <p>
        Here is the debug information you asked for. Please send this to the plugin maintainers if they ask for it.
      </p>
      <div
        className="c4 c5"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="c6 c1"
  >
    <div
      className="c7"
      color="red"
    >
      <span>
        Oh dear, an error!
      </span>
      <a
        href="#"
        onClick={[Function]}
      >
        Close
      </a>
    </div>
    <div
      className="c3"
    >
      <p>
        Looks like we've run into a problem. You could help the developers out and 
        <a
          className="send-button"
          href="#"
          onClick={[Function]}
        >
          send the error report.
        </a>
        <br />
        <a
          href="#"
          onClick={[Function]}
        >
          What does this mean?
        </a>
      </p>
      <div
        className="c4 c5"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="c8"
  >
    <div
      className="c9"
    >
      <div
        className="c10"
      />
    </div>
    <span>
      Loading
    </span>
  </div>
  <div
    className="c8"
  >
    <div
      className="c11"
    >
      <div />
      <button
        className="c12"
        onClick={[Function]}
      >
        OK
      </button>
    </div>
  </div>
  <div
    className="c13 c8"
  >
    <p />
  </div>
  <div
    className="c14"
  >
    <p>
      foobar
    </p>
  </div>
</div>
`;
